cocotb/triggers.py

```diff
@@ -735,3 +735,44 @@ def _wait(self):
         for _ in range(self.num_cycles):
             yield trigger
         raise ReturnValue(self)
+
+
+@decorators.coroutine
+def with_timeout(triggers, timeout_time, timeout_unit=None):
+    """
+    Waits on triggers, throws an exception if it waits longer than the given time
+
+    Can be used with any expression that would otherwise appear to the right of a
+    `yield` (or `await` in Python 3) statement.
+
+    Usage:
+
+    .. code-block:: python
+
+        yield with_timeout(coro, 100, 'ns')
+        yield with_timeout([coro, event.wait(), ...], 100, 'ns')
+
+    Args:
+        triggers (cocotb_yield_expr):
+            Any expression that could be right of a yield statement in cocotb
+        timeout_time (int):
+            Time duration
+        timeout_unit (str, optional):
+            Units of duration, accepts any values that :class`cocotb.triggers.Timer` does
+
+    Returns:
+        First trigger that completed if timeout did not occur.
+
+    Raises:
+        :exc:`SimTimeoutError`: If timeout occurs
+    """
+
+    if not isinstance(triggers, list):
+        triggers = [triggers]
+
+    timeout_timer = cocotb.triggers.Timer(timeout_time, timeout_unit)
+    res = yield [timeout_timer,] + triggers
```

Comment:
I'm curious if you can drop the `if` above and just use:
```
res = yield cocotb.triggers.First(timeout_trigger, triggers)
```

Reply:
`timeout_timer` is a trigger and `triggers` is either a single trigger or a list of triggers. `First` expects an arglist of triggers. That would work if you splatted `triggers`, but you would have to condition that on it being a `list` first. You can't really get around normalizing it.

You could force `timeout_time` and `timeout_unit` to be keyword-only and then take an arglist, but end up adding a check to `timeout_time` to force the user to give a value. That might be cleaner though.